### PR TITLE
[experiment] Add generative search for llm

### PR DESCRIPTION
**Pull request by w5688414**
### Description

# Training

+ single gpu
```
python run.py --do_train \
              --model_name_or_path bigscience/bloomz-7b1-mt \
              --output_dir ./checkpoints \
              --train_data ./data/toy_finetune_data.jsonl \
              --overwrite_output_dir \
              --fine_tune_type bitfit \
              --sentence_pooling_method weighted_mean \
              --num_train_epochs 10 \
              --per_device_train_batch_size 4 \
```

+ multi gpus

```
python -m paddle.distributed.launch --gpus "4,5,6,7" run.py \
        --do_train \
        --model_name_or_path bigscience/bloomz-7b1-mt  \
        --num_train_epochs 10 --per_device_train_batch_size 4 \
        --evaluation_strategy no \
        --save_steps 30 \
        --passage_max_len 300 \
        --bf16 --fp16_opt_level O2 --tensor_parallel_degree 4 \
        --logging_steps 50 --output_dir outputs \
        --sentence_pooling_method weighted_mean \
        --fine_tune_type bitfit \
        --overwrite_output_dir \
        --train_data ./data/toy_finetune_data.jsonl \
        --is_batch_negative True \
        --save_total_limit 10 

python merge_tp_params.py --model_name_or_path outputs/checkpoint-30
```

### CODE REVIEW

File: examples/semantic_indexing/llm/modeling.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            self.process_rank = dist.get_rank()
+            self.world_size = dist.get_world_size()
+
+    def sentence_embedding(self, hidden_state, mask):
+        if self.sentence_pooling_method == "mean":
+            s = paddle.sum(hidden_state * mask.unsqueeze(-1), axis=1)
+            d = mask.sum(axis=1, keepdim=True)
+            return s / d
+        elif self.sentence_pooling_method == "cls":
+            return hidden_state[:, 0]
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 这两个method都不太合理。
大模型基本没有cls, 可以排除。
mean的话，对于causal模型来说等于严重偏向靠前的token, 而少考虑靠后的token, 不太合理
如果不会用到，是不是可以去调了

File: examples/semantic_indexing/llm/modeling.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        elif self.sentence_pooling_method == "weighted_mean":
+            # Use weighted mean to compute similarity for decoder only LLMs
+            # refer to https://github.com/Muennighoff/sgpt/blob/9728de441b1dd2e638a8a64e1c83f77716f47d9a/biencoder/beir/beir_dense_retriever.py#L258
+            weights = (
+                paddle.arange(start=1, end=hidden_state.shape[1] + 1)
+                .unsqueeze(0)
+                .unsqueeze(-1)
+                .expand(hidden_state.shape)
+            )
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py:214]: 所以这段的逻辑是越后的token weight越高？比如如果seq length是1024, 那么第一个token的embedding权重是1, 最后1个token是1024, 也就是第一个token的1000倍?

File: examples/semantic_indexing/llm/modeling.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            sum_mask = paddle.sum(input_mask_expanded * weights, axis=1)
+            embedding = sum_embeddings / sum_mask
+            return embedding
+
+    def encode(self, features):
+        if features is None:
+            return None
+        psg_out = self.model(**features, return_dict=True)
+        p_reps = self.sentence_embedding(psg_out.last_hidden_state, features["attention_mask"])
+        if self.normlized:
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: normalized

File: examples/semantic_indexing/llm/data.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            return_tensors="pd",
+        )
+        d_collated = self.tokenizer(
+            passage,
+            padding=True,
+            truncation=True,
+            return_attention_mask=True,
+            max_length=self.passage_max_len,
+            return_tensors="pd",
+        )
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/data.py]: collator中调用tokenizer会很慢，建议挪到数据预处理的部分，因为collator是需要在训练的时候on the fly做的

File: examples/semantic_indexing/llm/data.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            self.dataset = datasets.load_dataset("json", data_files=args.train_data, split="train")
+
+        self.tokenizer = tokenizer
+        self.args = args
+        self.total_len = len(self.dataset)
+
+    def __len__(self):
+        return self.total_len
+
+    def __getitem__(self, item) -> Tuple[BatchEncoding, List[BatchEncoding]]:
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/data.py]: 你这个没有tokenizer, 哪里得到的BatchEncoding?

File: examples/semantic_indexing/llm/modeling.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            p_reps = paddle.nn.functional.normalize(p_reps, axis=-1)
+        return p_reps.contiguous()
+
+    def compute_similarity(self, q_reps, p_reps):
+        if len(p_reps.shape) == 2:
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 把q_reps, p_reps的shape都列一下。
为啥p_reps会有3维的情况呢？不应该都是[batch, hidden]吗

File: examples/semantic_indexing/llm/modeling.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        dist.all_gather(all_tensors, t)
+
+        all_tensors[self.process_rank] = t
+        all_tensors = paddle.concat(all_tensors, axis=0)
+
+        return all_tensors
+
+    def save(self, output_dir: str):
+        state_dict = self.model.state_dict()
+        state_dict = type(state_dict)({k: v.clone().cpu() for k, v in state_dict.items()})
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 没看懂这行

File: examples/semantic_indexing/llm/utils.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        # Save a trained model and configuration using `save_pretrained()`.
+        # They can then be reloaded using `from_pretrained()`
+        if not hasattr(self.model, "save"):
+            raise NotImplementedError(f"MODEL {self.model.__class__.__name__} " f"does not support save interface")
+        else:
+            self.model.save(output_dir)
+        if self.tokenizer is not None and self.is_world_process_zero():
+            self.tokenizer.save_pretrained(output_dir)
+
+    def compute_loss(self, model, inputs, return_outputs=False):
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/utils.py:19]: 为啥这个需要override呢？直接按照第一个model返回值是loss不行吗？

File: examples/semantic_indexing/llm/modeling.py (hunk 3 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py]: 已修改

File: examples/semantic_indexing/llm/modeling.py (hunk 2 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py:214]: image

对的，根据论文里面来的，最后会除以一个分母

File: examples/semantic_indexing/llm/data.py (hunk 5 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/data.py]: 已删除

File: examples/semantic_indexing/llm/modeling.py (hunk 2 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py:79]: image
对的，这个根据论文里面来的，最后会有一个分母。

image

File: examples/semantic_indexing/llm/modeling.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    ):
+        query = inputs["query"]
+        passage = inputs["passage"]
+        q_reps = self.encode(query)
+        p_reps = self.encode(passage)
+
+        if self.training:
+            if self.negatives_cross_device:
+                q_reps = self._dist_gather_tensor(q_reps)
+                p_reps = self._dist_gather_tensor(p_reps)
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py:120]: 这个_dist_gather_tensor仅在sharding和dp场景下有意义，对于tp和pp需要直接raise error

File: examples/semantic_indexing/llm/modeling.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+            return embedding
+
+    def encode(self, features):
+        if features is None:
+            return None
+        psg_out = self.bloom(**features, return_dict=True)
+        p_reps = self.sentence_embedding(psg_out.last_hidden_state, features["attention_mask"])
+        if self.normalized:
+            p_reps = paddle.nn.functional.normalize(p_reps, axis=-1)
+        return p_reps.contiguous()
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: paddle有contiguous？

File: examples/semantic_indexing/llm/modeling.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+            input_mask_expanded = mask.unsqueeze(-1).expand(hidden_state.shape)
+            # bs, seq_len, hidden_dim -> bs, hidden_dim
+            sum_embeddings = paddle.sum(hidden_state * input_mask_expanded * weights, axis=1, dtype="float32")
+            sum_mask = paddle.sum(input_mask_expanded * weights, axis=1)
+            embedding = sum_embeddings / sum_mask
+            return embedding
+
+    def encode(self, features):
+        if features is None:
+            return None
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py:95]: 什么时候有这种none的case?

File: examples/semantic_indexing/llm/modeling.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+            # else:
+            self.process_rank = dist.get_rank()
+            self.world_size = dist.get_world_size()
+
+    def sentence_embedding(self, hidden_state, mask):
+        if self.sentence_pooling_method == "weighted_mean":
+            # Use weighted mean to compute similarity for decoder only LLMs
+            # refer to https://github.com/Muennighoff/sgpt/blob/9728de441b1dd2e638a8a64e1c83f77716f47d9a/biencoder/beir/beir_dense_retriever.py#L258
+            weights = (
+                paddle.arange(start=1, end=hidden_state.shape[1] + 1)
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 这里的paddle.arange可以在__init__里面提前预制好，不用每次都生成

File: examples/semantic_indexing/llm/modeling.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+            if not dist.is_initialized():
+                raise ValueError("Distributed training has not been initialized for representation all gather.")
+            #     logger.info("Run in a single GPU, set negatives_cross_device=False")
+            #     self.negatives_cross_device = False
+            # else:
+            self.process_rank = dist.get_rank()
+            self.world_size = dist.get_world_size()
+
+    def sentence_embedding(self, hidden_state, mask):
+        if self.sentence_pooling_method == "weighted_mean":
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py:80]: 只有一个就不用if了

File: examples/semantic_indexing/llm/modeling.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+        if self.training:
+            if self.negatives_cross_device:
+                q_reps = self._dist_gather_tensor(q_reps)
+                p_reps = self._dist_gather_tensor(p_reps)
+            if self.is_batch_negative:
+                # In batch negatives
+                scores = self.compute_similarity(q_reps, p_reps)
+                # Substract margin from all positive samples cosine_sim()
+                margin_diag = paddle.full(
+                    shape=[q_reps.shape[0]], fill_value=self.margin, dtype=paddle.get_default_dtype()
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 不要用get_default_dtype, 直接dtype=q_reps.dtype

File: examples/semantic_indexing/llm/modeling.py (hunk 12 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py]: 咋生成，创建一个tensor放着？但每个batch的序列长度不一样咋办？

File: examples/semantic_indexing/llm/modeling.py (hunk 13 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py:80]: 还有一个last hidden 的策略没有实现

File: examples/semantic_indexing/llm/modeling.py (hunk 9 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py:120]: 这是多卡的cross batch negatives策略，如果是ddp，这个策略可以用到。

File: examples/semantic_indexing/llm/data.py (hunk 4 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/data.py]: 已修改

File: examples/semantic_indexing/llm/modeling.py (hunk 6 above)
Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py]: 已修改

File: examples/semantic_indexing/llm/modeling.py (hunk 12 above)
Comment (sijunhe) [reply examples/semantic_indexing/llm/modeling.py]: 预制一个模型支持最大长度的arange tensor, 根据输入长度slice前n个element就好

File: examples/semantic_indexing/llm/modeling.py (hunk 9 above)
Comment (sijunhe) [reply examples/semantic_indexing/llm/modeling.py:120]: 我知道，我的意思是如果tp和pp你是不能用这个策略的，需要raise error

File: examples/semantic_indexing/llm/modeling.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+                # Scale cosine to ease training converge
+                scores = scores / self.temperature
+                target = paddle.arange(0, q_reps.shape[0], dtype="int64")
+                loss = self.compute_loss(scores, target)
+            else:
+                scores = self.compute_similarity(q_reps, p_reps)
+                scores = scores / self.temperature
+                scores = scores.reshape([q_reps.shape[0], -1])
+
+                target = paddle.arange(scores.shape[0], dtype="int64")
```
Comment (sijunhe) [inline examples/semantic_indexing/llm/modeling.py]: 这里的2个arange和上面说的同理，都可以直接复用变量

Comment (w5688414) [reply examples/semantic_indexing/llm/modeling.py]: 已修改
